File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteSingleScheduleRecord.java
package com.krystianwsul.checkme.firebase.records;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.krystianwsul.checkme.firebase.json.ScheduleWrapper;
import com.krystianwsul.checkme.firebase.json.SingleScheduleJson;

import junit.framework.Assert;

public class RemoteSingleScheduleRecord extends RemoteScheduleRecord {
    RemoteSingleScheduleRecord(@NonNull String id, @NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(id, remoteTaskRecord, scheduleWrapper);
    }

    RemoteSingleScheduleRecord(@NonNull RemoteTaskRecord remoteTaskRecord, @NonNull ScheduleWrapper scheduleWrapper) {
        super(remoteTaskRecord, scheduleWrapper);
    }

    @NonNull
    private SingleScheduleJson getSingleScheduleJson() {
        SingleScheduleJson singleScheduleJson = getCreateObject().singleScheduleJson;
        Assert.assertTrue(singleScheduleJson != null);

        return singleScheduleJson;
    }

    @Override
    public long getStartTime() {
        return getSingleScheduleJson().getStartTime();
    }

    @Nullable
    @Override
    public Long getEndTime() {
        return getSingleScheduleJson().getEndTime();
    }

    public int getYear() {
        return getSingleScheduleJson().getYear();
    }

    public int getMonth() {
        return getSingleScheduleJson().getMonth();
    }

    public int getDay() {
        return getSingleScheduleJson().getDay();
    }

    @Nullable
    public String getCustomTimeId() {
        return getSingleScheduleJson().getCustomTimeId();
    }

    @Nullable
    public Integer getHour() {
        return getSingleScheduleJson().getHour();
    }

    @Nullable
    public Integer getMinute() {
        return getSingleScheduleJson().getMinute();
    }

    public void setEndTime(long endTime) {
        Assert.assertTrue(getEndTime() == null);

        getSingleScheduleJson().setEndTime(endTime);
        addValue(getKey() + "/singleScheduleJson/endTime", endTime);
    }
}

File: app/src/main/java/com/krystianwsul/checkme/firebase/records/RemoteSingleScheduleRecord.kt
package com.krystianwsul.checkme.firebase.records

import com.krystianwsul.checkme.firebase.json.ScheduleWrapper


class RemoteSingleScheduleRecord : RemoteScheduleRecord {

    private val singleScheduleJson by lazy { createObject.singleScheduleJson!! }

    override val startTime by lazy { singleScheduleJson.startTime }

    override var endTime
        get() = singleScheduleJson.endTime
        set(value) {
            if (value == singleScheduleJson.endTime)
                return

            singleScheduleJson.endTime = value
            addValue("$key/singleScheduleJson/endTime", value)
        }

    val year by lazy { singleScheduleJson.year }

    val month by lazy { singleScheduleJson.month }

    val day by lazy { singleScheduleJson.day }

    override val customTimeId by lazy { singleScheduleJson.customTimeId }

    val hour by lazy { singleScheduleJson.hour }

    val minute by lazy { singleScheduleJson.minute }

    constructor(id: String, remoteTaskRecord: RemoteTaskRecord, scheduleWrapper: ScheduleWrapper) : super(id, remoteTaskRecord, scheduleWrapper)

    constructor(remoteTaskRecord: RemoteTaskRecord, scheduleWrapper: ScheduleWrapper) : super(remoteTaskRecord, scheduleWrapper)

    override fun deleteFromParent() = check(remoteTaskRecord.remoteSingleScheduleRecords.remove(id) == this)
}
